### PR TITLE
Ctctraders 1908 -  917 response 

### PR DESCRIPTION
**Pull request by kenneth-rayner**

### CODE REVIEW

File: app/connectors/ArrivalMovementConnector.scala [hunk 1]
```diff
@@ -16,17 +16,25 @@
...
-import uk.gov.hmrc.http.{HeaderCarrier, HttpReads, HttpReadsTry}
+import uk.gov.hmrc.http.HttpReads.is2xx
+import uk.gov.hmrc.http.{HeaderCarrier, HttpReads, HttpReadsTry, HttpResponse}
 import uk.gov.hmrc.play.bootstrap.http.HttpClient
 
 import scala.concurrent.{ExecutionContext, Future}
+import scala.xml.NodeSeq
 
 class ArrivalMovementConnector @Inject()(config: FrontendAppConfig, http: HttpClient, ws: WSClient)(implicit ec: ExecutionContext) extends HttpReadsTry {
+  val logger: Logger = Logger(getClass)
```
Comment (amaalali) [inline app/connectors/ArrivalMovementConnector.scala:36]: To make adding logging easier, we've adopted [this pattern](https://github.com/hmrc/declare-transit-movement-arrival-frontend/blob/master/app/logging/Logging.scala), that way the loggers have a predictable pattern. I think that's worth adding to this repo

File: app/controllers/arrival/XmlNegativeAcknowledgementController.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    with I18nSupport {
+
+  def onPageLoad(arrivalId: ArrivalId): Action[AnyContent] = identify.async {
+    implicit request =>
+      arrivalMessageService.getXMLSubmissionNegativeAcknowledgementMessage(arrivalId).flatMap {
+        case Some(rejectionMessage) =>
+          val json = Json.obj("nctsEnquiries" -> frontendAppConfig.nctsEnquiriesUrl, "functionalError" -> rejectionMessage.error)
+
+          renderer.render("xmlNegativeAcknowledgement.njk", json).map(Ok(_))
+        case _ => Future.successful(Redirect(controllers.routes.TechnicalDifficultiesController.onPageLoad()))
```
Comment (amaalali) [inline app/controllers/arrival/XmlNegativeAcknowledgementController.scala:50]: Fo this we should return a 500 and render the technical difficulties page instead

File: app/models/ErrorType.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  case object ValueNotSupported extends GenericError(14)
+  case object NotSupportedPosition extends GenericError(15)
+  case object InvalidDecimal extends GenericError(19)
+  case object DuplicateDetected extends GenericError(26)
+  case object TooManyRepetitions extends GenericError(35)
+  case object InvalidTypeCharacters extends GenericError(37)
+  case object MissingDigit extends GenericError(38)
+  case object ElementTooLong extends GenericError(39)
+  case object ElementTooShort extends GenericError(40)
+  case class UnknownErrorCode(override val code: Int) extends GenericError(code)
```
Comment (amaalali) [inline app/models/ErrorType.scala:43]: Are these error codes specific to IE917 or do they have a generic meaning across messages like the IE008?

File: app/models/arrival/XMLSubmissionNegativeAcknowledgementMessage.scala [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+final case class XMLSubmissionNegativeAcknowledgementMessage(movementReferenceNumber: String, error: FunctionalError)
+
+object XMLSubmissionNegativeAcknowledgementMessage {
+
+  implicit val writes: OWrites[XMLSubmissionNegativeAcknowledgementMessage] = Json.writes[XMLSubmissionNegativeAcknowledgementMessage]
+
+  implicit val xmlReader: XmlReader[XMLSubmissionNegativeAcknowledgementMessage] = (
+    (__ \ "HEAHEA" \ "DocNumHEA5").read[String],
+    (__ \ "FUNERRER1").read[FunctionalError]
```
Comment (amaalali) [inline app/models/arrival/XMLSubmissionNegativeAcknowledgementMessage.scala:33]: Can there be zero or multiple errors?

File: test/controllers/arrival/XmlNegativeAcknowledgementControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+      val application = applicationBuilder(userAnswers = Some(emptyUserAnswers))
+        .overrides(inject.bind[ArrivalMessageService].toInstance(mockArrivalMessageService))
+        .build()
+
+      val request = FakeRequest(GET, routes.XmlNegativeAcknowledgementController.onPageLoad(arrivalId).url)
+
+      val result = route(application, request).value
+
+      status(result) mustEqual SEE_OTHER
```
Comment (amaalali) [inline test/controllers/arrival/XmlNegativeAcknowledgementControllerSpec.scala:93]: Should be 500, see comment in controller

**Review by amaalali — CHANGES_REQUESTED** [review]

File: app/controllers/arrival/XmlNegativeAcknowledgementController.scala (hunk 2 above)
Comment (ashwinirudrappa) [reply app/controllers/arrival/XmlNegativeAcknowledgementController.scala:50]: @amaalali As it was decided in one of our meetings to create a new ticket to address the mentioned issue do we need to implement the changes as a part of this PR

File: app/models/ErrorType.scala (hunk 3 above)
Comment (ashwinirudrappa) [reply app/models/ErrorType.scala:43]: @amaalali It is generic across messages like IE008 but not too strict if any other error code comes we still display it on the UI so the user can contact CCTO with those details

File: app/models/arrival/XMLSubmissionNegativeAcknowledgementMessage.scala (hunk 4 above)
Comment (ashwinirudrappa) [reply app/models/arrival/XMLSubmissionNegativeAcknowledgementMessage.scala:33]: @amaalali Its a mandatory object, not a sequence

**Comment by ashwinirudrappa** [comment]
test this please

**Comment by ashwinirudrappa** [comment]
test this please

**Review by amaalali — APPROVED** [review]

**Pull request merged**